porch/webhooks/main.go

```diff
@@ -0,0 +1,357 @@
+		return nil, err
+	}
+
+	return admissionReviewRequest, nil
+}
+
+func validateDeletion(w http.ResponseWriter, r *http.Request) {
+	klog.Infoln("received request to validate deletion")
+
+	admissionReviewRequest, err := admissionReviewFromRequest(r,
+		serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer())
+	if err != nil {
+		errMsg := fmt.Sprintf("error getting admission review from request: %v", err)
+		writeErr(errMsg, &w)
+		return
+	}
+
+	// Verify that we have a packagerevision resource
+	pkgRevGVK := metav1.GroupVersionResource{Group: "porch.kpt.dev", Version: "v1alpha1", Resource: "packagerevisions"}
+	if admissionReviewRequest.Request.Resource != pkgRevGVK {
+		errMsg := fmt.Sprintf("did not receive pkgRev, got %s", admissionReviewRequest.Request.Resource.Resource)
+		writeErr(errMsg, &w)
+		return
+	}
+
+	// Decode the package rev from the AdmissionReview.
+	cfg, err := config.GetConfig()
+	if err != nil {
+		klog.Errorf("could not get config: %s", err.Error())
+	}
+	porchClient, err := porch.CreateClient(cfg)
+	if err != nil {
+		klog.Errorf("could not get porch client: %s", err.Error())
+	}
+	pr := v1alpha1.PackageRevision{}
+	if err := porchClient.Get(context.Background(), client.ObjectKey{
+		Namespace: admissionReviewRequest.Request.Namespace,
+		Name:      admissionReviewRequest.Request.Name,
+	}, &pr); err != nil {
+		klog.Errorf("could not get package revision: %s", err.Error())
+	}
+
+	admissionResponse := &admissionv1.AdmissionResponse{}
+	if pr.Spec.DeletionProposed {
+		admissionResponse.Allowed = true
+		admissionResponse.Result = &metav1.Status{
+			Status:  "Success",
+			Message: fmt.Sprintf("Successfully deleted package revision %q", pr.Name),
+		}
+	} else {
+		admissionResponse.Allowed = false
+		admissionResponse.Result = &metav1.Status{
+			Status:  "Failure",
+			Message: fmt.Sprintf("failed to delete package revision %q: spec.deletionProposed not set to 'true'", pr.Name),
+			Reason:  "PackageRevisions must be proposed for deletion by setting spec.deletionProposed to 'true' prior to deletion.",
+		}
+	}
+
+	// Construct the response, which is just another AdmissionReview.
+	var admissionReviewResponse admissionv1.AdmissionReview
+	admissionReviewResponse.Response = admissionResponse
+	admissionReviewResponse.SetGroupVersionKind(admissionReviewRequest.GroupVersionKind())
+	admissionReviewResponse.Response.UID = admissionReviewRequest.Request.UID
+
+	resp, err := json.Marshal(admissionReviewResponse)
+	if err != nil {
+		errMsg := fmt.Sprintf("error marshalling response json: %v", err)
+		writeErr(errMsg, &w)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(resp)
+}
+
+func writeErr(errMsg string, w *http.ResponseWriter) {
+	klog.Errorf(errMsg)
+	(*w).WriteHeader(500)
+	if _, err := (*w).Write([]byte(errMsg)); err != nil {
+		klog.Errorf("could not write error message: %v", err)
+	}
+}
+
+func main() {
+	err := setupWebhooks(context.Background())
```

Comment:
Is there any meaningful unit tests we can write for this?

Reply:
I'll see if I can refactor some parts of it a bit to make it more unit-testable.

Reply:
I added unit tests for the cert-setup, and for error cases of the webhook server handler function.